### PR TITLE
[RF] Correctly implement additional dataset feature in HistFactory

### LINKED ISSUE #10538: AdditionalData is not exported to XML
- [x] Checked for duplicates

### Describe the bug

I tried to create "additional" datasets for one RooFit sample. They do not show up in the XML file (and it seems also not in the root file of the model). When discussing this with @cburgard , we found that the `fAdditionalData` member of `Channel` [is supposed to hold additional datasets to be added to the XML](https://github.com/root-project/root/blob/c8d49336b666355553879aea9448f7e9588ecb5e/roofit/histfactory/inc/RooStats/HistFactory/Channel.h#L96), but [is never used in `Channel::XML](https://github.com/root-project/root/blob/c8d49336b666355553879aea9448f7e9588ecb5e/roofit/histfactory/src/Channel.cxx#L105).

### Expected behavior

The additional datasets also get exported to XML/rootfile/etc.
### To Reproduce

Add an additional data sample to a workspace, save it.

I have modified one of the python examples so that it should reproduce this issue.
It was first discovered in our analysis framework.
[histfactoryxmlexport.py.txt](https://github.com/root-project/root/files/8641363/histfactoryxmlexport.py.txt)

### Setup

Both with ROOT 6.20/06 from `AnalysisBase,21.2.204` and ROOT 6.24/06 from [LCG_101_ATLAS_20](https://lcginfo.cern.ch/release/101_ATLAS_20/)

### PR DESCRIPTION
**Pull request by guitargeek**
In HistFactory, the presence of a `Channel::GetAdditionalDatas` function
hinted to the possibility of defining additional datases for a
HistFactory model, besides the nominal observations dataset named
`"obsData"`.

When reading a user-generated HistFactory XML, the additional datasets
were in fact read into the `Channel::fAdditionalData` member, and
corresponding RooDataSets were created in the per-channel proto
workspaces.

However, when defining the Measurement object in C++ and dumping the XML
via `PrintXML`, the additional datasets were not considered. They were
also not considered when merging the datasets in the per-channel proto
workspaces into the simultaneous dataset. This commit suggests to
implement this correctly.

Now, one can define additional datasets as follows in the XML:
```xml
<Data HistoName="data" InputFile="data/example.root" HistoPath="" Name="addData"/>
```
If there is no `Name` tag, the RooDataSet in the workspace will use the
`HistoName` as its name.

In the touched code, some cleaning of commented out code and general code modernization is also suggested.

Closes #10538.

### CODE REVIEW

**Comment by jndrf** [comment]
Hi @guitargeek , did you have time to look into this -- or maybe one of the reviewers? I don't have access to a mac, so I can't help troubleshooting these build failures.

**Comment by guitargeek** [comment]
Hi, thanks for asking! All is okay here, the test failures are unrelated to RooFit, and we are just waiting for the review by @lmoneta who was busy with other things last week. But this will come maybe today or tomorrow and then we merge this.

**Review by lmoneta — APPROVED** [review]
Thank you Jonas for this improvement!

**Pull request merged**
